mujoco_py/builder.py

```diff
@@ -292,6 +304,45 @@ def __exit__(self, type, value, traceback):
         cymj.set_warning_callback(self.prev_user_warning)
 
 
+def build_generic_fn(function_string, userdata_fields=[]):
+    '''
+    Builds a callback function with a `mjfGeneric` signature.
+    TODO: document userdata_fields
+    TODO: note important to remember these are just #defines
+    TODO: have debug flag/envvar which prints out source (or doesn't delete it)
+    TODO: document requirement that function have name `generic`
+    TODO: simple example
+    TODO: example iterating over contacts
+    TODO: example that defines other functions and calls them
+    TODO: tons of docs here
+    TODO: note possible future improvements, e.g. ability to link in external libs
+    TODO: Docurment MUJOCO_PY_DEBUG_FN_BUILDER=1 to keep files
+    '''
+    assert isinstance(userdata_fields, list), 'userdata_fields must be list'
+    ffibuilder = FFI()
+    ffibuilder.cdef('extern uintptr_t generic_fn;')
+    # TODO: time library building and note how long it should take
+    name = '_generic_fn_' + ''.join(choice(ascii_lowercase) for _ in range(15))
+    source_string = '#include <mujoco.h>\n'
+    # Add defines for each userdata to make setting them easier
+    for i, fieldname in enumerate(userdata_fields):
+        source_string += '#define {} d->userdata[{}]\n'.format(fieldname, i)
+    source_string += function_string
+    source_string += '\nuintptr_t generic_fn = (uintptr_t) generic;'
+    ffibuilder.set_source(name, source_string,
+                          include_dirs=[join(mjpro_path, 'include')],
+                          library_dirs=[join(mjpro_path, 'bin')],
+                          libraries=['mujoco150'])
+    library_path = ffibuilder.compile(verbose=True)
+    fixed_library_path = manually_link_libraries(mjpro_path, library_path)
+    module = load_dynamic_ext(name, fixed_library_path)
+    # Now that the module is loaded into memory, we can actually delete it
+    if not os.environ.get('MUJOCO_PY_DEBUG_FN_BUILDER', False):
+        for f in glob.glob(name + '*'):
+            os.remove(f)
+    return module.lib.generic_fn
```

Comment:
This is called "generic" because it's signature matches the `mjfGeneric` type defined here: https://github.com/openai/mujoco-py/blob/c5f60322467ec8ecc0db64c3e18a4da762c27e45/mujoco_py/pxd/mjdata.pxd#L288

(and also in the mjdata header)